### DISCUSSION TITLE
Issues with solid element - shell element bending comparison

### DISCUSSION THREAD

**Discussion by Edward-Eth** [idaholab/moose#21501] [Q&A Modules: Solid mechanics] [answered]
I have been attempting to compare bending performance of shell and solid elements in MOOSE (and comparing against FEMAP in addition) but have been having issues with my solid element model.

Both models are 20mm x 20mm x 1mm plates fixed on the x = 0 (left) edge and subjected to a 2N per unit area/length von neumann load on the right edge in the Z direction.

The shell model agrees closely with the result I get from Femap-Nastran, but the result from the solid model is much lower. When the out of plane load is replaced with a load in the x direction, the resulting displacements are very similar for both models which indicates the issue is more with how bending is being handled by the solid elements? The solid element results from femap-nastran are very comparable to the shell element results from both programs. I have implemented the same test using two different load application methods (this Neumann approach but also a NodalKernal method) with the same difference in displacement observed, indicating this is not a load application issue.

When higher degree solid elements are used, the discrepancy reduces, however Femap is able to produce good results with linear elements at the same resolution as the MOOSE model, so something is going wrong it would seem.

Shell Models, MOOSE and femap, with X direction stress (shell underside) plotted as contour. These images show that the displacements agree to a good tolerance level (6.11mm for Femap and 6.08 for MOOSE) but the stress magnitudes differ significantly.
<img width="1475" alt="Screenshot 2022-07-07 at 11 16 30" src="https://user-images.githubusercontent.com/64844573/177750816-19db449d-c0d8-4d1a-9a12-6d7cdeb9cbca.png">
<img width="1258" alt="Screenshot 2022-07-07 at 11 11 58" src="https://user-images.githubusercontent.com/64844573/177750863-c8471345-d982-49cc-bead-89e6d2ae962f.png">

Solid Models: MOOSE and femap, with von mises stress plotted as contour. These images show that while the femap solid model agrees closely with both shell models, the moose solid model is generating incorrect results, with a displacement of 2.39 instead of the expected ~6. Additionally, the stress values are much lower in the MOOSE model.
<img width="1472" alt="Screenshot 2022-07-07 at 11 17 00" src="https://user-images.githubusercontent.com/64844573/177751326-09940d06-86ce-4a00-9f51-13b4e2e36710.png">

<img width="1257" alt="Screenshot 2022-07-07 at 11 07 18" src="https://user-images.githubusercontent.com/64844573/177751345-341aaabb-876a-4478-be54-0ebcade3059c.png">

<details>
  <summary>Shell Element Model</summary>

```
# Shell element version of a shell vs solid bending comparison.
# The model is a cantilevered plate, 20mm x 20mm x 1mm.
# The left edge (x = 0) is fixed in X, Z and rotation about Y.
# The front edge is fixed in y.
# The right side of the plate is subject to a 2N per unit area load in the Z direction.

[Mesh]
  [Generator]
    type = GeneratedMeshGenerator
    dim = 2
    xmax = 20
    ymax = 20
    nx = 10
    ny = 10
  []
  [SampleXCenterline]
    type = ParsedGenerateSideset
    input = Generator
    new_sideset_name = Centerline
    combinatorial_geometry = 'y=10'
  []
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
  [rot_x]
  []
  [rot_y]
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Kernels]
  [solid_disp_x]
    type = ADStressDivergenceShell
    block = 0
    component = 0
    variable = disp_x
    through_thickness_order = SECOND
  []
  [solid_disp_y]
    type = ADStressDivergenceShell
    block = 0
    component = 1
    variable = disp_y
    through_thickness_order = SECOND
  []
  [solid_disp_z]
    type = ADStressDivergenceShell
    block = 0
    component = 2
    variable = disp_z
    through_thickness_order = SECOND
  []
  [solid_rot_x]
    type = ADStressDivergenceShell
    block = 0
    component = 3
    variable = rot_x
    through_thickness_order = SECOND
  []
  [solid_rot_y]
    type = ADStressDivergenceShell
    block = 0
    component = 4
    variable = rot_y
    through_thickness_order = SECOND
  []
[]

[Materials]
  [stress]
    type = ADComputeShellStress
    through_thickness_order = SECOND
  []
  [elasticity]
    type = ADComputeIsotropicElasticityTensorShell
    youngs_modulus = 10000
    poissons_ratio = 0.3
    through_thickness_order = SECOND
  []
  [strain]
    type = ADComputeIncrementalShellStrain
    displacements = 'disp_x disp_y disp_z'
    rotations = 'rot_x rot_y'
    thickness = 1
    through_thickness_order = THIRD
  []
[]

[BCs]
  [Fixed_X]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_x
  []
  [Fixed_Y]
    type = ADDirichletBC
    boundary = Centerline
    value = 0
    variable = disp_y
  []
  [Fixed_Z]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_z
  []
  [Fixed_rot_y]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = rot_y
  []
  [Load]
    type = ADNeumannBC
    boundary = right
    value = 2
    variable = disp_z
  []
[]

[AuxVariables]
  [stress_00]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [stress_00]
    type = RankTwoAux
    variable = stress_00
    rank_two_tensor = global_stress_t_points_0
    index_i = 0
    index_j = 0
    execute_on = TIMESTEP_END
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = false
  []
[]

[Executioner]
  type = Steady
  automatic_scaling = TRUE
  solve_type = 'NEWTON'
  line_search = 'default'
  petsc_options_iname = '-pc_type  -snes_linesearch_type -pc_factor_shift_type -snes_max_it  '
                        '-ksp_max_it'
  petsc_options_value = 'lu        basic                 NONZERO               20            30'
[]

[Postprocessors]
  [Point_Displacement_X]
    type = PointValue
    point = '20 10 0'
    variable = disp_x
  []
  [Point_Displacement_Y]
    type = PointValue
    point = '20 10 0'
    variable = disp_y
  []
  [Point_Displacement_Z]
    type = PointValue
    point = '20 10 0'
    variable = disp_z
  []
[]

[Outputs]
  [Exodus]
    type = Exodus
    file_base = 'IsotropicShellBending'
  []
[]
```
</details>

<details>
<summary>Solid Element Model</summary>

```
# Solid element version of a shell vs solid bending stress comparison.
# The model is a cantilevered plate, 20mm x 20mm x 1mm.
# There are 4 elements through thickness.
# The left edge (x = 0) is fixed in X and Z, this also prevents rotation about Y.
# The front edge (y = 0) is fixed in Y.
# The right side of the plate is subject to a 2N per unit area load in the Z direction.

[Mesh]
  [Generator]
    type = GeneratedMeshGenerator
    dim = 3
    xmax = 20
    ymax = 20
    zmin = -0.5
    zmax = 0.5
    nx = 10
    ny = 10
    nz = 4
  []
  [SampleXCenterline]
    type = ParsedGenerateSideset
    input = Generator
    new_sideset_name = Centerline
    combinatorial_geometry = 'y=10'
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    incremental = true
    add_variables = true
    use_finite_deform_jacobian = true
    volumetric_locking_correction = false
    use_automatic_differentiation = true
    generate_output = vonmises_stress
  []
[]

[Materials]
  [StandardElasticityTensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 10000
    poissons_ratio = 0.3
  []
  [Stress]
    type = ADComputeLinearElasticStress
  []
[]

[BCs]
  [Fixed_X]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_x
  []
  [Fixed_Y]
    type = ADDirichletBC
    boundary = Centerline
    value = 0
    variable = disp_y
  []
  [Fixed_Z]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_z
  []
  [Load]
    type = ADNeumannBC
    boundary = right
    value = 2
    variable = disp_z
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Steady
  automatic_scaling = True
  solve_type = 'NEWTON'
  line_search = 'default'
  petsc_options_iname = '-pc_type  -snes_linesearch_type -pc_factor_shift_type -snes_max_it  '
                        '-ksp_max_it'
  petsc_options_value = 'lu        basic                 NONZERO               20            30'
  nl_abs_tol = 1e-50
  # nl_rel_tol = 1e-10
[]

[Postprocessors]
  [Point_Displacement_X]
    type = PointValue
    point = '20 10 0'
    variable = disp_x
  []
  [Point_Displacement_Y]
    type = PointValue
    point = '20 10 0'
    variable = disp_y
  []
  [Point_Displacement_Z]
    type = PointValue
    point = '20 10 0'
    variable = disp_z
  []
[]

[Outputs]
  [Exodus]
    type = Exodus
    file_base = 'IsotropicSolidBending'
  []
[]

```
</details>

**Comment by GiudGiud** [comment]
Hello

Are you using this branch?
https://github.com/idaholab/moose/pull/21385
It fixed an issue with those elements iirc

Guillaume

**Comment by Edward-Eth** [comment]
The model I am having an issue with is the solid element model, the shells are behaving correctly here.

**Reply by GiudGiud** [reply]
oh my bad, read too fast. 

@dschwen @jiangwen84  who works on solid elements?

**Comment by Edward-Eth** [comment]
The plot thickens: When I use hex20 or hex28 elements, the solid model displacement is a lot closer to the shell element displacement (and the 'correct' value) however it is still significantly different, and femap is able to produce correct solid element results using linear solid elements (hex8). This should mean that more linear elements through thickness would improve the result, however it does not, no quantity of linear elements through thickness seems to change the result which to me implies there's something going wrong here.

**Comment by GiudGiud** [comment]
I think the load is applied differently
Is the centerline supposed to traverse the width of the beam? Or is it supposed to just be on top of it?

**Reply by Edward-Eth** [reply]
The centerline passes through the middle of the beam (in the y direction), through the full thickness, as shown in this image. This is just used to enforce a y=0 constraint which is purely a stability feature, there are no loads in the y direction.
<img width="1084" alt="Screenshot 2022-07-07 at 16 39 33" src="https://user-images.githubusercontent.com/64844573/177814642-1e03c588-53f6-4726-b951-63b348e520d3.png">

**Reply by GiudGiud** [reply]
ok not that one then
this one maybe?
```
  [Load]
    type = ADNeumannBC
    boundary = right
    value = 2
    variable = disp_z
  []
```
you have the same thing for shell elements,  but the area of application of this flux is different

**Comment by neuphris** [comment]
Regarding the difference in stress one of the possible reason can be that stresses are calculated at each quadrature point, the output from the MOOSE that you are seeing is the average of the stresses in the qps of the element unless you have used "selected_qp = " in the AuxKernel. I am not sure what stress NASTRAN gives as output.

**Comment by hugary1995** [comment]
I took a brief look at the NASTRAN theory. It seems to me that NASTRAN uses hyperelasticity for large deformation. If you can confirm this, then the results won't be reproducible using the tensor_mechanics module unfortunately.

But I am not saying MOOSE cannot do hyperelasticity, it's just that the tensor_mechanics module cannot.

**Reply by hugary1995** [reply]
Better yet, if you tell me which [hyperelastic material](http://www2.me.rochester.edu/courses/ME204/nx_help/index.html#uid:id1340937) you are using in NASTRAN, I can give you a MOOSE-equivalent to try.

**Reply by hugary1995** [reply]
Also, let's try not to get to the topic of which large deformation formulation is better... I'm a hyperelasticity person, but I still use hypoelasticity (tensor_mechanics) very often.

**Comment by Edward-Eth** [comment, marked as answer]
Going to go through these responses one by one because I don't think any of them actual address the issue I'm seeing here.

1) @GiudGiud 
> ok not that one then this one maybe?
> 
> ```
>   [Load]
>     type = ADNeumannBC
>     boundary = right
>     value = 2
>     variable = disp_z
>   []
> ```
> 
> you have the same thing for shell elements, but the area of application of this flux is different

This was a concern of mine originally, but when I apply the Neumann BC to disp_x instead, the resulting in plane displacements are identical, which to me implies that applying to a line edge is equivalent to applying to a face of unit width and the same length as the line. To verify that the method of force application is not the issue, I ran models using a constant rate nodal kernel instead, with the load applied to the nodes on the loaded edge of the shell, and the mid-thickness nodes of the loaded edge in the solid. The images below show the nodes used, showing that each model applies a load to the same number of nodes (11) in the same locations.

<img width="1397" alt="Screenshot 2022-07-12 at 09 07 31" src="https://user-images.githubusercontent.com/64844573/178441674-d785b9e9-6fad-4596-b448-ea003551b875.png">

<img width="1351" alt="Screenshot 2022-07-12 at 09 10 24" src="https://user-images.githubusercontent.com/64844573/178442218-65a24ddc-6a24-4eac-a3f2-d1cac935c233.png">



<details>
  <summary>Shell Element Model Using Nodal Kernel</summary>

```
# Shell element version of a shell vs solid bending comparison.
# The model is a cantilevered plate, 20mm x 20mm x 1mm.
# The left edge (x = 0) is fixed in X, Z and rotation about Y.
# The front edge is fixed in y.
# The right side of the plate is subject to a 2N per unit area load in the Z direction.

[Mesh]
  [Generator]
    type = GeneratedMeshGenerator
    dim = 2
    xmax = 20
    ymax = 20
    nx = 10
    ny = 10
  []
  [SampleXCenterline]
    type = ParsedGenerateSideset
    input = Generator
    new_sideset_name = Centerline
    combinatorial_geometry = 'y=10'
  []
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
  [rot_x]
  []
  [rot_y]
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Kernels]
  [solid_disp_x]
    type = ADStressDivergenceShell
    block = 0
    component = 0
    variable = disp_x
    through_thickness_order = FOURTH
  []
  [solid_disp_y]
    type = ADStressDivergenceShell
    block = 0
    component = 1
    variable = disp_y
    through_thickness_order = FOURTH
  []
  [solid_disp_z]
    type = ADStressDivergenceShell
    block = 0
    component = 2
    variable = disp_z
    through_thickness_order = FOURTH
  []
  [solid_rot_x]
    type = ADStressDivergenceShell
    block = 0
    component = 3
    variable = rot_x
    through_thickness_order = FOURTH
  []
  [solid_rot_y]
    type = ADStressDivergenceShell
    block = 0
    component = 4
    variable = rot_y
    through_thickness_order = FOURTH
  []
[]

[Materials]
  [stress]
    type = ADComputeShellStress
    through_thickness_order = FOURTH
  []
  [elasticity]
    type = ADComputeIsotropicElasticityTensorShell
    youngs_modulus = 10000
    poissons_ratio = 0.3
    through_thickness_order = FOURTH
  []
  [strain]
    type = ADComputeIncrementalShellStrain
    displacements = 'disp_x disp_y disp_z'
    rotations = 'rot_x rot_y'
    thickness = 1
    through_thickness_order = FOURTH
  []
[]

[BCs]
  [Fixed_X]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_x
  []
  [Fixed_Y]
    type = ADDirichletBC
    boundary = Centerline
    value = 0
    variable = disp_y
  []
  [Fixed_Z]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_z
  []
  [Fixed_rot_y]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = rot_y
  []
  [Load_Z]
    type = ADNeumannBC
    boundary = right
    value = 2
    variable = disp_z
  []
  # [Load_X]
  #   type = ADNeumannBC
  #   boundary = right
  #   value = 2
  #   variable = disp_x
  # []
[]

[NodalKernels]
  [Z_Load]
    type = ConstantRate
    rate = 2
    variable = disp_z
    boundary = right
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = false
  []
[]

[Executioner]
  type = Steady
  automatic_scaling = TRUE
  solve_type = 'NEWTON'
  line_search = 'default'
  petsc_options_iname = '-pc_type  -snes_linesearch_type -pc_factor_shift_type -snes_max_it  '
                        '-ksp_max_it'
  petsc_options_value = 'lu        basic                 NONZERO               20            30'
[]

[Postprocessors]
  [Point_Displacement_X]
    type = PointValue
    point = '20 10 0'
    variable = disp_x
  []
  [Point_Displacement_Y]
    type = PointValue
    point = '20 10 0'
    variable = disp_y
  []
  [Point_Displacement_Z]
    type = PointValue
    point = '20 10 0'
    variable = disp_z
  []
[]

[Outputs]
  [Exodus]
    type = Exodus
    file_base = 'IsotropicShellBending'
  []
[]

```
</details>

<details>
  <summary>Solid Element Model Using Nodal Kernel</summary>

```
# Solid element version of a shell vs solid bending stress comparison.
# The model is a cantilevered plate, 20mm x 20mm x 1mm.
# There are 4 elements through thickness.
# The left edge (x = 0) is fixed in X and Z, this also prevents rotation about Y.
# The front edge (y = 0) is fixed in Y.
# The right side of the plate is subject to a 2N per unit area load in the Z direction.

[Mesh]
  [Generator]
    type = GeneratedMeshGenerator
    dim = 3
    xmax = 20
    ymax = 20
    zmin = -0.5
    zmax = 0.5
    nx = 10
    ny = 10
    nz = 4
  []
  [SampleXCenterline]
    type = ParsedGenerateSideset
    input = Generator
    new_sideset_name = Centerline
    combinatorial_geometry = 'y=10'
  []
  [Right_Edge_Center_Nodes]
    type = BoundingBoxNodeSetGenerator
    input = SampleXCenterline
    new_boundary = Right_Mid_Nodes
    bottom_left = '19.99 -0.01 -0.01'
    top_right = '20.01 20.01 0.01'
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    incremental = true
    add_variables = true
    use_finite_deform_jacobian = true
    volumetric_locking_correction = false
    use_automatic_differentiation = true
    generate_output = vonmises_stress
  []
[]

[Materials]
  [StandardElasticityTensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 10000
    poissons_ratio = 0.3
  []
  [Stress]
    type = ADComputeLinearElasticStress
  []
[]

[BCs]
  [Fixed_X]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_x
  []
  [Fixed_Y]
    type = ADDirichletBC
    boundary = Centerline
    value = 0
    variable = disp_y
  []
  [Fixed_Z]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_z
  []
  # [Load]
  #   type = ADNeumannBC
  #   boundary = right
  #   value = 2
  #   variable = disp_z
  # []
[]

[NodalKernels]
  [Z_Load]
    type = ConstantRate
    rate = 2
    variable = disp_z
    boundary = Right_Mid_Nodes
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Steady
  automatic_scaling = True
  solve_type = 'NEWTON'
  line_search = 'default'
  petsc_options_iname = '-pc_type  -snes_linesearch_type -pc_factor_shift_type -snes_max_it  '
                        '-ksp_max_it'
  petsc_options_value = 'lu        basic                 NONZERO               20            30'
  nl_abs_tol = 1e-50
  # nl_rel_tol = 1e-10
[]

[Postprocessors]
  [Point_Displacement_X]
    type = PointValue
    point = '20 10 0'
    variable = disp_x
  []
  [Point_Displacement_Y]
    type = PointValue
    point = '20 10 0'
    variable = disp_y
  []
  [Point_Displacement_Z]
    type = PointValue
    point = '20 10 0'
    variable = disp_z
  []
[]

[Outputs]
  [Exodus]
    type = Exodus
    file_base = 'IsotropicSolidBending'
  []
[]
```
</details>

In this case, the disparity when loaded in the z direction is still present (and is in fact greater), but again, when loaded in x the displacements are equal, indicating that the issue does not lie in the magnitude of applied forces.

2) @neuphris 
> Regarding the difference in stress one of the possible reason can be that stresses are calculated at each quadrature point, the output from the MOOSE that you are seeing is the average of the stresses in the qps of the element unless you have used "selected_qp = " in the AuxKernel. I am not sure what stress NASTRAN gives as output.

Yeah I'm not comparing apples to apples here necessarily, so I would disregard this for now, I'm likely making an error in comparison.

3) @hugary1995 
> I took a brief look at the NASTRAN theory. It seems to me that NASTRAN uses hyperelasticity for large deformation. If you can confirm this, then the results won't be reproducible using the tensor_mechanics module unfortunately.
> 
> But I am not saying MOOSE cannot do hyperelasticity, it's just that the tensor_mechanics module cannot.

I'm not using an hyperelasticity in the Nastran model. I can only assume you got that impression from the use of finite strain in the solid input file. That was an edit I made to attempt to get the results to agree, but it actually makes a negligble difference. With strain = FINITE, I get a displacement in the solid model of 2.39mm, with strain = SMALL I get a displacement of 2.43mm, both compared to the correct result of around 6mm. The material property used in Nastran is a simple linear elastic the same as implemented in moose using the isotropic elasticity tensor.

4) 
> The plot thickens: When I use hex20 or hex28 elements, the solid model displacement is a lot closer to the shell element displacement (and the 'correct' value) however it is still significantly different, and femap is able to produce correct solid element results using linear solid elements (hex8). This should mean that more linear elements through thickness would improve the result, however it does not, no quantity of linear elements through thickness seems to change the result which to me implies there's something going wrong here.

This comment seems to have slipped under the radar somewhat so I'll mention it again here. Using this input file (NOTE: Small strain so hyperelasticty is not a factor here): 

<details>
  <summary>Solid Element Model With Neumann Load</summary>

```
# Solid element version of a shell vs solid bending stress comparison.
# The model is a cantilevered plate, 20mm x 20mm x 1mm.
# There are 4 elements through thickness.
# The left edge (x = 0) is fixed in X and Z, this also prevents rotation about Y.
# The front edge (y = 0) is fixed in Y.
# The right side of the plate is subject to a 2N per unit area load in the Z direction.

[Mesh]
  [Generator]
    type = GeneratedMeshGenerator
    dim = 3
    xmax = 20
    ymax = 20
    zmin = -0.5
    zmax = 0.5
    nx = 10
    ny = 10
    nz = 4
    elem_type = HEX8
  []
  [SampleXCenterline]
    type = ParsedGenerateSideset
    input = Generator
    new_sideset_name = Centerline
    combinatorial_geometry = 'y=10'
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = SMALL
    incremental = true
    add_variables = true
    use_finite_deform_jacobian = true
    volumetric_locking_correction = false
    use_automatic_differentiation = true
    generate_output = vonmises_stress
  []
[]

[Materials]
  [StandardElasticityTensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 10000
    poissons_ratio = 0.3
  []
  [Stress]
    type = ADComputeLinearElasticStress
  []
[]

[BCs]
  [Fixed_X]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_x
  []
  [Fixed_Y]
    type = ADDirichletBC
    boundary = Centerline
    value = 0
    variable = disp_y
  []
  [Fixed_Z]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_z
  []
  [Load]
    type = ADNeumannBC
    boundary = right
    value = 2
    variable = disp_z
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Steady
  automatic_scaling = True
  solve_type = 'NEWTON'
  line_search = 'default'
  petsc_options_iname = '-pc_type  -snes_linesearch_type -pc_factor_shift_type -snes_max_it  '
                        '-ksp_max_it'
  petsc_options_value = 'lu        basic                 NONZERO               20            30'
  nl_abs_tol = 1e-50
  # nl_rel_tol = 1e-10
[]

[Postprocessors]
  [Point_Displacement_X]
    type = PointValue
    point = '20 10 0'
    variable = disp_x
  []
  [Point_Displacement_Y]
    type = PointValue
    point = '20 10 0'
    variable = disp_y
  []
  [Point_Displacement_Z]
    type = PointValue
    point = '20 10 0'
    variable = disp_z
  []
[]

[Outputs]
  [Exodus]
    type = Exodus
    file_base = 'IsotropicSolidBending'
  []
[]
```
</details>

When run as is, I get a displacement of 2.43mm. If I change Mesh/Generator/elem_type to HEX20 I get a displacement of 6.05mm. If I change it to HEX27 I get a displacement of 6.05mm, all of these compared to the correct result of around 6mm. This indicates to me that the HEX8 element type in MOOSE is performing very inaccurately in bending loads.

If instead of changing the element type (therefore using HEX8 again), I change the number of elements through the thickness of the model, the variable most commonly associated with bending accuracy using solid elements, I get the following: Baseline 4 elements: 2.43mm, 6 elements: 2.43mm, 8 elements: 2.44mm, 10 elements: 2.44mm. 

This seems to indicate that no amount of HEX8 elements can capture bending correctly, whereas in Nastran, the 4 elements through thickness gives near perfect results, and in fact 2 elements through thickness gives almost identical results as well.

I'm quite concerned that I've stumbled upon another implementation bug here (as with the shell element stress bug I discovered) so would be keen for someone with more knowledge of the back end of tensor_mechanics to investigate.

**Reply by hugary1995** [reply]
> I'm quite concerned that I've stumbled upon another implementation bug here (as with the shell element stress bug I discovered) so would be keen for someone with more knowledge of the back end of tensor_mechanics to investigate.

Instead of figuring out whether there's a bug on the tensor_mechanics side, we really need more information on the Nastran side. I can guarantee that in tensor_mechanics if you do linear elasticity, it really is _just_ linear elastcity... It's not a matter of element implementation, and it's _impossible_ to get it wrong. However, Nastran might be doing some sort of stablization behind the scenes.

When you try to benchmark software B to an existing model in software A, and especially you are opening this discussion in B's forum, the first logical step is to provide as much information as you can regarding the model in A, instead of asking what's wrong with B. Hope this makes sense to you.

**Reply by Edward-Eth** [reply]
Regardless of what Nastran is doing I don't see how changing the element type from HEX8 to HEX20 should lead to a change of almost 3x in displacement?

**Reply by hugary1995** [reply]
To be more clear, all I can gather from your previous posts about the Nastran elasticity model is it is some sort of solid element with linear elasticity. We will need a Nastran expert to tell us more about its formulation so that we can do our best to set up the same thing in tensor_mechanics.

**Reply by hugary1995** [reply]
> Regardless of what Nastran is doing I don't see how changing the element type from HEX8 to HEX20 should lead to a change of almost 3x in displacement?

The geometry and arrangement of elements you have is subject to shear locking, among other stability issues.

**Reply by Edward-Eth** [reply]
Okay, apologies. I've done some digging and you're absolutely correct, Nastran uses 'Reduced Shear Integration with bubble functions' to reduce shear locking. I had enabled volumetric locking in my input and in my mind gotten that swapped around with shear locking, leading me to think that MOOSE had a similar feature that 'wasn't working' when I now realise that it's obviously a different feature. How this reduced shear integration method works seems to be a bit of a secret sauce method. An obvious solution to my issue is the use of higher order elements, so I'll go ahead with that approach.

**Reply by hugary1995** [reply]
No problem. Glad you figured it out. And yes, bubble function is the first guess I would make as well. We don't have bubble function enriched Lagrange shape functions in libMesh yet. This is actually the second time the need for bubble function comes up, the first one being #17569.
